File: 1/main.go
package main

import (
	"math"
	"os"
	"slices"
	"strconv"
	"strings"
)

func main() {
	file, err := os.ReadFile("input.txt")
	if err != nil {
		panic(err)
	}

	lines := strings.Split(strings.TrimSpace(string(file)), "\n")
	firstList, secondList := getTwoLists(lines)

	// Part 1
	var difference []int
	for index, firstElement := range firstList {
		secondElement := secondList[index]
		diff := int(math.Abs(float64(secondElement - firstElement)))
		difference = append(difference, diff)
	}

	println(sum(difference))

	// Part 2
	var similarityScores []int
	for _, n := range firstList {
		score := similarityScore(n, secondList)
		similarityScores = append(similarityScores, score)
	}

	println(sum(similarityScores))
}

func sum(list []int) int {
	var sum int
	for _, n := range list {
		sum += n
	}
	return sum
}

func similarityScore(num int, list []int) int {
	var score int
	for _, n := range list {
		if n == num {
			score++
		}
	}
	return score * num
}

func getTwoLists(lines []string) ([]int, []int) {
	var list1, list2 []int
	for _, line := range lines {
		words := strings.Fields(line)
		firstWord := words[0]
		secondWord := words[1]

		n, err := strconv.Atoi(firstWord)
		if err != nil {
			panic(err)
		}
		m, err := strconv.Atoi(secondWord)
		if err != nil {
			panic(err)
		}

		list1 = append(list1, n)
		list2 = append(list2, m)
	}

	slices.Sort(list1)
	slices.Sort(list2)

	return list1, list2
}

File: 1/Main.hs
module Main where

import           Data.List
import           Data.Maybe
import           System.IO

parseLine :: String -> Maybe (Int, Int)
parseLine str =
    case words str of
      [a, b] -> Just (read a, read b)
      _      -> Nothing

linesToTwoLists :: [String] -> ([Int], [Int])
linesToTwoLists str = (sort l1, sort l2)
  where
    (l1, l2) = unzip . mapMaybe parseLine $ str

similarityScore :: [Int] -> Int -> Int
similarityScore xs n = n * (length . filter (== n) $ xs)

main :: IO ()
main = do
    handle <- openFile "input.txt" ReadMode
    lines' <- lines <$> hGetContents handle
    let (l1, l2) = linesToTwoLists lines'
    -- Part 1
    print $ sum $ map abs $ zipWith (-) l1 l2
    -- Part 2
    print $ sum $ map (similarityScore l1) l2
